### PR TITLE
ARROW-8996: [C++] AVX2/AVX512 runtime support for aggregate sum kernel

### PR DESCRIPTION
**Pull request by frankdjx**
1. Add AVX2/AVX512 path.
2. Add build support to append the compiler flag for SIMD code file.
3. Register the SIMD version at runtime as the CPU feature, also add avx2/avx512 detect.
4. Also add a test case to cover the overflow on the data types which need a right shift
    on the Accumulator type.

Signed-off-by: Frank Du <frank.du@intel.com>

### CODE REVIEW

**Comment by jianxind** [comment]
Benchmark data:

Before:
```
SumKernelFloat/32768/0         2.96 us         2.96 us       236912 bytes_per_second=10.3227G/s null_percent=0 size=32.768k
SumKernelFloat/32768/1         4.88 us         4.88 us       143527 bytes_per_second=6.25439G/s null_percent=1 size=32.768k
SumKernelFloat/32768/10        5.13 us         5.13 us       136839 bytes_per_second=5.95117G/s null_percent=10 size=32.768k
SumKernelFloat/32768/50        7.82 us         7.81 us        87129 bytes_per_second=3.9054G/s null_percent=50 size=32.768k
SumKernelDouble/32768/0        1.97 us         1.97 us       356786 bytes_per_second=15.4906G/s null_percent=0 size=32.768k
SumKernelDouble/32768/1        2.11 us         2.11 us       331511 bytes_per_second=14.4975G/s null_percent=1 size=32.768k
SumKernelDouble/32768/10       2.39 us         2.38 us       291292 bytes_per_second=12.7966G/s null_percent=10 size=32.768k
SumKernelDouble/32768/50       2.60 us         2.60 us       268800 bytes_per_second=11.7462G/s null_percent=50 size=32.768k
SumKernelInt8/32768/0          11.7 us         11.7 us        59926 bytes_per_second=2.61569G/s null_percent=0 size=32.768k
SumKernelInt8/32768/1          11.0 us         10.9 us        63640 bytes_per_second=2.78831G/s null_percent=1 size=32.768k
SumKernelInt8/32768/10         14.8 us         14.8 us        46573 bytes_per_second=2.05848G/s null_percent=10 size=32.768k
SumKernelInt8/32768/50         14.6 us         14.6 us        47840 bytes_per_second=2.08905G/s null_percent=50 size=32.768k
SumKernelInt16/32768/0         7.06 us         7.06 us        99354 bytes_per_second=4.3245G/s null_percent=0 size=32.768k
SumKernelInt16/32768/1         4.76 us         4.75 us       147305 bytes_per_second=6.41928G/s null_percent=1 size=32.768k
SumKernelInt16/32768/10        5.64 us         5.63 us       122737 bytes_per_second=5.42002G/s null_percent=10 size=32.768k
SumKernelInt16/32768/50        6.71 us         6.70 us       104192 bytes_per_second=4.55206G/s null_percent=50 size=32.768k
SumKernelInt32/32768/0         3.92 us         3.92 us       178798 bytes_per_second=7.79042G/s null_percent=0 size=32.768k
SumKernelInt32/32768/1         3.27 us         3.27 us       214296 bytes_per_second=9.332G/s null_percent=1 size=32.768k
SumKernelInt32/32768/10        3.41 us         3.40 us       204944 bytes_per_second=8.9683G/s null_percent=10 size=32.768k
SumKernelInt32/32768/50        3.69 us         3.69 us       190248 bytes_per_second=8.27705G/s null_percent=50 size=32.768k
SumKernelInt64/32768/0         1.92 us         1.91 us       368662 bytes_per_second=15.9508G/s null_percent=0 size=32.768k
SumKernelInt64/32768/1         2.05 us         2.05 us       340168 bytes_per_second=14.8684G/s null_percent=1 size=32.768k
SumKernelInt64/32768/10        2.16 us         2.16 us       323585 bytes_per_second=14.1164G/s null_percent=10 size=32.768k
SumKernelInt64/32768/50        2.41 us         2.41 us       291073 bytes_per_second=12.6873G/s null_percent=50 size=32.768k
```

After:
```
SumKernelFloat/32768/0         2.27 us         2.27 us       307928 bytes_per_second=13.438G/s null_percent=0 size=32.768k
SumKernelFloat/32768/1         4.59 us         4.59 us       152827 bytes_per_second=6.6508G/s null_percent=1 size=32.768k
SumKernelFloat/32768/10        5.30 us         5.29 us       132106 bytes_per_second=5.76658G/s null_percent=10 size=32.768k
SumKernelFloat/32768/50        5.80 us         5.80 us       114378 bytes_per_second=5.26584G/s null_percent=50 size=32.768k
SumKernelDouble/32768/0        1.42 us         1.42 us       494426 bytes_per_second=21.5265G/s null_percent=0 size=32.768k
SumKernelDouble/32768/1        2.12 us         2.12 us       330890 bytes_per_second=14.4268G/s null_percent=1 size=32.768k
SumKernelDouble/32768/10       2.44 us         2.43 us       286310 bytes_per_second=12.5441G/s null_percent=10 size=32.768k
SumKernelDouble/32768/50       2.72 us         2.71 us       257105 bytes_per_second=11.2507G/s null_percent=50 size=32.768k
SumKernelInt8/32768/0          5.35 us         5.34 us       130751 bytes_per_second=5.71315G/s null_percent=0 size=32.768k
SumKernelInt8/32768/1          9.80 us         9.79 us        71384 bytes_per_second=3.11589G/s null_percent=1 size=32.768k
SumKernelInt8/32768/10         13.9 us         13.9 us        49729 bytes_per_second=2.19116G/s null_percent=10 size=32.768k
SumKernelInt8/32768/50         12.5 us         12.5 us        55929 bytes_per_second=2.43479G/s null_percent=50 size=32.768k
SumKernelInt16/32768/0         3.20 us         3.19 us       218923 bytes_per_second=9.55594G/s null_percent=0 size=32.768k
SumKernelInt16/32768/1         5.31 us         5.31 us       131394 bytes_per_second=5.75174G/s null_percent=1 size=32.768k
SumKernelInt16/32768/10        6.20 us         6.19 us       113037 bytes_per_second=4.92965G/s null_percent=10 size=32.768k
SumKernelInt16/32768/50        7.25 us         7.24 us        96604 bytes_per_second=4.21535G/s null_percent=50 size=32.768k
SumKernelInt32/32768/0         2.18 us         2.18 us       321572 bytes_per_second=14.0037G/s null_percent=0 size=32.768k
SumKernelInt32/32768/1         3.32 us         3.32 us       209911 bytes_per_second=9.18857G/s null_percent=1 size=32.768k
SumKernelInt32/32768/10        3.59 us         3.58 us       195106 bytes_per_second=8.51472G/s null_percent=10 size=32.768k
SumKernelInt32/32768/50        3.83 us         3.82 us       182739 bytes_per_second=7.98056G/s null_percent=50 size=32.768k
SumKernelInt64/32768/0         1.37 us         1.37 us       514237 bytes_per_second=22.3564G/s null_percent=0 size=32.768k
SumKernelInt64/32768/1         2.09 us         2.09 us       333678 bytes_per_second=14.5962G/s null_percent=1 size=32.768k
SumKernelInt64/32768/10        2.18 us         2.18 us       320094 bytes_per_second=13.9904G/s null_percent=10 size=32.768k
SumKernelInt64/32768/50        2.41 us         2.40 us       289766 bytes_per_second=12.6907G/s null_percent=50 size=32.768k
```

All dense part of data types has some improvements , ex Double jump to 21.5265G/s from 15.4906G/s.

The sparse parts I will look into later as it need some additional to remove the invalid value before passing to the SIMD add operations, it need some shuffle op to replace the invalid value to zero.

Also the dense part can be speed up again if using AVX2/AVX512 which is a later job also.

**Comment by jianxind** [comment]
@wesm Can you help to review the approach for runtime kernel path?

For the build part, I introduced some compiler option(GCC push_options or clang attribute push) to help build the SIMD code without a common SIMD flag defined in CXX flags.

For the SSE implementation, currently only dense part is implemented with intrinsic, I need more work for how to remove the invalid values.

**Comment by jianxind** [comment]
@ursabot benchmark --suite-filter=arrow-compute-aggregate-benchmark

**Comment by wesm** [comment]
I'll try to review this in the next couple days

**Comment by wesm** [comment]
I'll look at this today

**Comment by pitrou** [comment]
This looks generally quite complicated. If we need 500 additional lines of code to micro-optimize the Sum kernel for a single SIMD instruction set (nevermind that we may also want versions for AVX2, Neon, SVE, and whatnot), things will quickly get out of hand.

If we want to go the way of per-kernel SIMD optimizations, it may be useful to investigate SIMD helper libraries (such as [libsimdpp](https://github.com/p12tic/libsimdpp), [xsimd](https://xsimd.readthedocs.io/en/latest/)...).

**Comment by fsaintjacques** [comment]
@ursabot benchmark --suite-filter=arrow-compute-aggregate-benchmark --cc=clang --cxx=clang++

**Comment by jianxind** [comment]
> This looks generally quite complicated. If we need 500 additional lines of code to micro-optimize the Sum kernel for a single SIMD instruction set (nevermind that we may also want versions for AVX2, Neon, SVE, and whatnot), things will quickly get out of hand.
> 
> If we want to go the way of per-kernel SIMD optimizations, it may be useful to investigate SIMD helper libraries (such as [libsimdpp](https://github.com/p12tic/libsimdpp), [xsimd](https://xsimd.readthedocs.io/en/latest/)...).

libsimdpp says it has a runtime dispatch support based on function basic. But I don't know if the quality of these libs is perfect enough to use and it was qualified in every spec, it will make the debug things more hard for me if there's any bug there. Another concern is the performance, will the wrapper introduce extra cost? And usually the SIMD code/parameters is designed carefully for each architecture to get best performance, I suspect there's a common code can fit all target.

**Comment by pitrou** [comment]
Certainly someone (or several people) would have to evaluate those libraries before deciding whether to integrate one. Preferably someone with an interest in SIMD optimizations (for example you and/or @cyb70289).

But saying we don't want to use a helper library because the code might be suboptimal performance-wise and it's better to rewrite it from scratch doesn't sound like good engineering practice to me. Daily Arrow maintenance is done by a small set of developers and we should refrain from creating too much burden.

**Comment by jianxind** [comment]
Just find a document(https://dl.acm.org/doi/pdf/10.1145/3178433.3178435 PAGE 7, Table 6: Comparison of various SIMD wrappers.) Some SIMD helper(simdpp/xsimd) has performance issue  at least on some workload.

Another thing is most SIMD helpers has no runtime support, it means we still has to build same code(if we can find a common code path for one function) many times on arrow itself for the runtime capacity.

And I'm working on the sparse part for aggregate sum recently, the data flow is total different for AVX2/AVX512. AVX512 has _mm512_mask_add_pd (__m512d src, __mmask8 k, __m512d a, __m512d b) support that it can SIMD add the results directly on the valid bit map. For AVX2, it has to use a lookup table mask with SIMD and operation to zero the invalid values before passing to SIMD add. The difference is applied to other future SIMD func also as all arrow data represented with valid bit map.

**Comment by pitrou** [comment]
> Just find a document(https://dl.acm.org/doi/pdf/10.1145/3178433.3178435 PAGE 7, Table 6: Comparison of various SIMD wrappers.) Some SIMD helper(simdpp/xsimd) has performance issue at least on some workload.

Interesting comparison, thank you. But that is a single workload and we don't know what the issue was precisely.

By the way, this is what the paper says about the approach you're advocating:

> The intrinsics or assembly approaches are non-portable, low-level solutions which target specific architectures. They offer maximum control to take advantage of instruction set specificities, and to fine tune register usage. However, it is quite difficult to develop and maintain a low-level code in the long run.

File: cpp/src/arrow/compute/kernels/aggregate_basic_sse.cc [hunk 1]
```diff
@@ -0,0 +1,355 @@
...
+    }                                                                                    \
+                                                                                         \
+    /* Add the values to result streams */                                               \
+    for (int64_t batch = 0; batch < num_batch; batch++) {                                \
+      for (int i = 0; i < kSseBatchStreams; i++) {                                       \
+        const auto src_simd =                                                            \
+            SimdLoadFn(&values[batch * kSseBatchSize + kSseStreamSize * i]);             \
+        results_simd[i] = SimdAddFn(src_simd, results_simd[i]);                          \
+      }                                                                                  \
+    }                                                                                    \
```
Comment (jedbrown) [inline cpp/src/arrow/compute/kernels/aggregate_basic_sse.cc:76]: Have you considered using OpenMP SIMD (i.e., `-fopenmp-simd`, which does not do anything with threads or require linking OpenMP)?
```c
double sum1(int64_t n, const double *values) {
    double sum = 0;
    #pragma omp simd reduction(+:sum)
    for (int64_t i = 0; i < n; i++)
        sum += values[i];
    return sum;
}
```
produces good vectorized code (probably excessively unrolled by clang)
```
        vaddpd  zmm0, zmm0, zmmword ptr [rsi + 8*rdx]
        vaddpd  zmm1, zmm1, zmmword ptr [rsi + 8*rdx + 64]
        vaddpd  zmm2, zmm2, zmmword ptr [rsi + 8*rdx + 128]
        vaddpd  zmm3, zmm3, zmmword ptr [rsi + 8*rdx + 192]
        vaddpd  zmm0, zmm0, zmmword ptr [rsi + 8*rdx + 256]
        vaddpd  zmm1, zmm1, zmmword ptr [rsi + 8*rdx + 320]
        vaddpd  zmm2, zmm2, zmmword ptr [rsi + 8*rdx + 384]
        vaddpd  zmm3, zmm3, zmmword ptr [rsi + 8*rdx + 448]
        vaddpd  zmm0, zmm0, zmmword ptr [rsi + 8*rdx + 512]
        vaddpd  zmm1, zmm1, zmmword ptr [rsi + 8*rdx + 576]
        vaddpd  zmm2, zmm2, zmmword ptr [rsi + 8*rdx + 640]
        vaddpd  zmm3, zmm3, zmmword ptr [rsi + 8*rdx + 704]
        vaddpd  zmm0, zmm0, zmmword ptr [rsi + 8*rdx + 768]
        vaddpd  zmm1, zmm1, zmmword ptr [rsi + 8*rdx + 832]
        vaddpd  zmm2, zmm2, zmmword ptr [rsi + 8*rdx + 896]
        vaddpd  zmm3, zmm3, zmmword ptr [rsi + 8*rdx + 960]
        .loc    1 6 5                   # ./example.c:6:5
        sub     rdx, -128
        add     rcx, -4
        jne     .LBB0_7
```
https://gcc.godbolt.org/z/6mD6oy

Comment (jianxind) [reply cpp/src/arrow/compute/kernels/aggregate_basic_sse.cc:76]: Thanks, the vectorize code by clang is really good. But seems the GCC part only use zmm0 which may limit the performance. This PR only aims on dense part, the final target is expanding the support to sparse part also, do you know if sparse can applied similar approach also? Example scalar code like below

  inline T MaskedValue(bool valid, T value) const { return valid ? value : 0; }
      for (size_t i = 0; i < 8; i++) {
        local.sum += MaskedValue(bits & (1U << i), values[i]);
      }
      local.count += BitUtil::kBytePopcount[bits];

Comment (jedbrown) [reply cpp/src/arrow/compute/kernels/aggregate_basic_sse.cc:76]: You can use GCC pragmas to force more unrolling, but AVX512 vectorization gets you a ways despite the adder instruction latency. (My experience has been that GCC does an excellent job with intrinsics, often better than ICC, and tends to favor smaller code at comparable optimization levels. Clang tends to be more aggressive than GCC at unrolling, though I've often seen that not pay off.)

The masked case is harder and I'm not aware of an idiomatic way to get GCC to vectorize it. We get some masked vectorized code with clang, though it isn't as clean as I'd hope for.
```c
double sum1(int64_t n, const unsigned char *mask, const double *values) {
    double sum = 0;
    #pragma omp simd reduction(+:sum) collapse(2)
    for (int64_t i = 0; i < n/8; i++) {
        for (int64_t j = 0; j < 8; j++) {
            sum += mask[i] & (1 << j) ? values[i*8+j] : 0;
        }
    }
    return sum;
}
```

https://gcc.godbolt.org/z/84anmr

**Comment by jianxind** [comment]
@ursabot benchmark --suite-filter=arrow-compute-aggregate-benchmark

**Comment by jianxind** [comment]
@ursabot benchmark --suite-filter=arrow-compute-aggregate-benchmark

**Comment by jianxind** [comment]
@ursabot benchmark --suite-filter=arrow-compute-aggregate-benchmark

**Comment by jianxind** [comment]
@ursabot benchmark --suite-filter=arrow-compute-aggregate-benchmark

**Comment by jianxind** [comment]
The numbers looks good for sparse part on the benchmark device also now, even though locally my device can get more better speed up thanks to bigger L2/L3 size. I don't know if it's a good direction considering compiler can do less things on the bitmap based. 

@emkornfield And others, thoughts?

**Comment by emkornfield** [comment]
@jianxind please see my e-mail.  I think we should try to split this PR up into runtime dispatch using slightly tweaked version of the current code without intrinsics.  I will take a closer look at what is being done for sparse, but if you aren't using BitBlockCounter we should incorporate that as a first step, so most sums can delegate to Dense

**Comment by emkornfield** [comment]
I think we also need a way of setting max runtime instruction set for runtime dispatch (apologies if there is one and I missed it)

**Comment by jianxind** [comment]
> I think we also need a way of setting max runtime instruction set for runtime dispatch (apologies if there is one and I missed it)

Thanks, currently no. But it can easily supported by hooking into CpuInfo::Init(),  tweak hardware_flags_ while checking the user ENV settings, something like "ARROW_RUNTIME_SIMD_LEVEL=none make unittest".

For split PR without intrinsics, I will try to work this. Now I used some compiler pragma(clang attribute push or GCC push_options) to support the intrinsic API, it's defined to NULL for MSVC as MSVC supports directly using intrinsic API without compiler flag. Need find a new method for new ways without intrinsic, maybe something like set_source_files_properties can help here.

**Comment by jianxind** [comment]
Close this one as reference of intrinsic. I will try to work out a new approach without intrinsic.

**Pull request closed without merging**